### ISSUE TITLE
Binary search for Test-PasswordQuality

### ISSUE THREAD

**Issue by magl8** [MichaelGrafnetter/DSInternals#79]
Would it be feasible to implement a binary search to improve the comparing speed?
As the password lists from HIBP are pre-sorted, this could be a promising approach.

**Comment by MichaelGrafnetter** [comment]
The search is actually done using a hash table, which contains all AD account password hashes. So the time required to lookup HIBP hashes only increases linearly with the size of the HIBP list. It is therefore faster than a binary search.
Do you have performance issues? Are you sure that they are caused by `Test-PasswordQuality` itself and not by `Get-ADDBAccount`/`Get-ADReplAccount`?
Asking Troy Hunt to publish a reduced HIBP file containing only passwords that meet AD complexity requirements would probably be the way to go. I am also working on replacing regular indices with containerized indices for AD DB account enumeration.

**Comment by magl8** [comment]
Extracting the hashes using Get-ADDBAccount takes just a few seconds. 
The bulk of time is spent using Test-PasswordQuality. 

Using the current version pwned-passwords-ntlm-ordered-by-hash-v4.txt (19'891 MBytes)
the Test-PasswordQuality function takes about one hour , regardless of the count of accounts/hashes to test.

Why build a hash table using all 550M hashes from the v4 file and not just for the hashes to be checked, which are  much less (at least for most of us admins out there)? 

Using an adapted version of Chris Dent's Binary Searching Script, the searching/comparison time for 2000 hashes against the same v4 file was less than 2 minutes, and this includes the conversion into Hashcat format using Get-ADDBAccount, reading the resulting file line-by-line and writing the results to a new file line-by-line. Here's a link to Chris' Source: 
[https://gist.github.com/indented-automation/9abd843e752b48ae7910314b878d594b#file-find-string-ps1](url)

**Comment by MichaelGrafnetter** [comment]
Thx for reply, @magl8. Well, the code is doing exactly what you described. It is only storing the hashes to be checked, not those 550M. It only took a couple of minutes on my PC when I tested it the last time. The file is then processed in a single pass, see (TestPasswordQualityCommand.cs)[https://github.com/MichaelGrafnetter/DSInternals/blob/master/Src/DSInternals.PowerShell/Commands/Misc/TestPasswordQualityCommand.cs]:
```CS
/// <summary>
/// Maps password hashes to lists of account names.
/// </summary>
private IDictionary<byte[], SortedSet<string>> hashToAccountMap;
...
// We need to cache NT hashes of all accounts for the Duplicate and Weak Password Tests.
this.hashToAccountMap = new Dictionary<byte[], SortedSet<string>>(PasswordDictionaryInitialCapacity, HashEqualityComparer.GetInstance());
...
private void TestWeakPasswordHash(byte[] weakHash)
{
    SortedSet<string> matchingAccounts;
    bool foundAccounts = this.hashToAccountMap.TryGetValue(weakHash, out matchingAccounts);
    if (foundAccounts)
    {
        this.result.WeakPassword.UnionWith(matchingAccounts);
    }
}
```
So I will have to run a profiler to check where the bottleneck is. It might either be the `Dictionary` or `StreamReader`. Or do you have an idea while looking at the code?

**Comment by magl8** [comment]
Well, I'm not really a programmer...

Looking at your code at a conceptional level, I could imagine that the use of SortedDictionary instead of IDictionary could speed up the searching process (because it inherently uses binary search).

Perhaps it would help to check for the existance of a hash with the ContainsKey method instead of trying to retrieve it in full with the TryGetValue method.

**Comment by magl8** [comment]
@MichaelGrafnetter:

Thank you for your patience.

Reading up on the subject, I found the problem of your code: You are building your own dictionary hash table, reading the whole 550M perfectly pre-sorted hashes into your own dictionary table. Here's the code snippet:

![image](https://user-images.githubusercontent.com/48722586/55280645-bca15e00-5328-11e9-8c40-83b9825bb10a.png)

I suggest you do NOT re-hash the whole pre-sorted file from HIBP. This is unnecessary because the file is already sorted and contains the desired information without the need for conversion. Instead, you can use the HIBP file to directly look up ("cherry-pick") the hash codes taken from the AD extract using a file-based binary searching function.  

This results in a searching time of 0.05 seconds per user (using the aforementioned PS code from Chris Dent). I would say that this is a massive improvement compared to a standard 1 hour waiting.

Of course, this approach only works if the pwned ntlm hash list file is pre-sorted by hash.
And looking up more than 72k user hashes at once would be the break even point in comparison to the current approach. For these users and for those requiring the use of unsorted hash lists, you could implement a command line parameter to enforce the building of your own dictionary table.

Do you think it would be feasible to implement a "direct lookup" alternative into DSInternals?

**Comment by magl8** [comment]
Here is the code I used for the file-based binary search:

```
function Find-String {
    <#
    .SYNOPSIS
        Find a matching string from an alphabetically sorted file.
    .DESCRIPTION
        Find-String is a specialised binary (half interval) searcher designed to find matches in sorted ASCII encoded text files.
        Usage example: $Pwned = find-string $TestHash pwned-passwords-ntlm-ordered-by-hash-v4.txt
    .NOTES
        Change log:
            11/08/2014 - Chris Dent - First release.
            03/30/2019 - Marius Gaudin - Adaption to haveibeenpwned.com sorted hash list files (both SHA1 and ntlm).
    #>

    [CmdletBinding()]
    [OutputType([Boolean])]
    param (
        # The string to find. The string is treated as a regular expression and must match the beginning of the line.
        [Parameter(Mandatory = $true)]
        [String]$String,

        # The name of the file to search.
        [Parameter(Mandatory = $true)]
        [ValidateScript( { Test-Path $_ } )]
        [String]$FileName
    )

    $FileName = $pscmdlet.GetUnresolvedProviderPathFromPSPath($FileName)
    $FileStream = New-Object IO.FileStream($FileName, [IO.FileMode]::Open)
    $BinaryReader = New-Object IO.BinaryReader($FileStream)
    $pwned=$false

    $Length = $BinaryReader.BaseStream.Length
    $Position = $Length / 2

    [Int64]$HalfInterval = $Length / 2
    $Position = $Length - $HalfInterval

    while ($Position -gt 1 -and $Position -lt $Length -and $Position -ne $LastPosition) {
        $LastPosition = $Position
        $HalfInterval = $HalfInterval / 2

        $BinaryReader.BaseStream.Seek($Position, [IO.SeekOrigin]::Begin) | Out-Null

        # Track back to the start of the line
        while ($true) {
            $Character = $BinaryReader.ReadByte()
            if ($BinaryReader.BaseStream.Position -eq 1) {
                $BinaryReader.BaseStream.Seek(-1, [IO.SeekOrigin]::Current) | Out-Null
                break
            } elseif ($Character -eq [Byte][Char]"`n") {
                break
            } else {
                $BinaryReader.BaseStream.Seek(-2, [IO.SeekOrigin]::Current) | Out-Null
            }
        }

        # Read the line
        $Characters = @()
        if ($BinaryReader.BaseStream.Position -lt $BinaryReader.BaseStream.Length) {
            do {
                $Characters += [Char][Int]$BinaryReader.ReadByte()
            } until ($Characters[-1] -eq [Char]"`n" -or $BinaryReader.BaseStream.Position -eq $BinaryReader.BaseStream.Length)
            $Line = (New-Object String (,[Char[]]$Characters)).split(":")[0].Trim()
         } else {
            # End of file
            $FileStream.Close()
            return $pwned
        }

        if ($Line -match "^$String") {
            # Close the file stream and return the match immediately
            $FileStream.Close()
            $pwned = $true
            return $pwned
        } elseif ($Line -lt $String) {
            $Position = $Position + $HalfInterval
        } elseif ($Line -gt $String) {
            $Position = $Position - $HalfInterval
        }
    }

    # Close the file stream if no matches are found.
    $FileStream.Close()
    return $pwned
}

```

**Comment by MichaelGrafnetter** [comment]
Thanks, @magl8, for the input. After doing some tests, I have to admit that it really takes too much time to process the file. Even the following simple command took more than 1 hour to finish:

```powershell
Measure-Command { Get-Content 'pwned-passwords-ntlm-ordered-by-hash-v4.7.txt' }
```
It would therefore appear that not reading the entire file and doing a binary search is the best option. I will thus add a new parameter called `-WeakPasswordHashesOrderedFile` and implement the binary search in one of the next DSInternals releases.

**Comment by MichaelGrafnetter** [comment]
Implemented in [DSInternals 3.4](https://github.com/MichaelGrafnetter/DSInternals/blob/master/Documentation/CHANGELOG.md#34---2019-04-23).
